### PR TITLE
feat: Limit Oauth access to not allow private repo and org data

### PR DESCRIPTION
**Pull request by bdougie**
## What type of PR is this? (check all applicable)

- [ ] ♻️ Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 👷 Optimization
- [ ] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

## Description
I have known for a while we needed to migrate this project from an OAuth app to a GitHub App, but due to bandwidth or other bugs I have not to look into this transition until today. 

I have already created the GitHub App, but need to address this without breaking access to the existing users. So here are the next steps. 

- [x] Handle the open-sauced-goals creation and repo App install using new GraphQL mutations.
- [x] Create a new error if the user has not enabled access to their username/open-sauced-goals repo. 

QA notes:
- If you try to test this today, note you will need to manually log out appending `/logout` to the URL.
- Still a WIP

## Related Tickets & Documents
fixes https://github.com/open-sauced/open-sauced/issues/696
fixes https://github.com/open-sauced/open-sauced/issues/689

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
TBD

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [x] 📜 readme
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Comment by bdougie** [comment]
## Quick update: 

I can confirm that login works on the deploy preview. I now need to enable specific interactions and permissions in the app. 

https://github.com/apps/open-sauced

## GitHub App Thoughts
The benefit of having the GitHub App is that I can prevent access to private organization data and repos. With this being a project to manage open source, that data is not needed.

I now need to discover the best way to enable this App install on the open-sauced-goals repo. I plan to work on that this evening. When I finish this, I will ping the triage team for help in the QA before merging.

**Comment by bdougie** [comment]
update: 

## Problem
The APP works fine as an OAuth app, but it asks for too much data. A GitHub (implement in this PR) ask for less data but fails while trying to use the same GraphQL queries (apiGraphQL.js). 

Screenshot 2020-06-24 13 22 15

## Limitation of GitHub Apps
It appears that GitHub Apps cannot access the following attributes:
- forks
- issues
- pullRequests
- stargazers

For those interested, I was able to mock everything [insomia](https://insomnia.rest/) thanks to my coworker today.

## Quick solution

I met with the OneGraph team and we discovered a workaround that would only be temporary until GitHub Apps are figured out, but I want to wait to implement until my meeting with the GitHub App team internally. 

Thanks for everyone's patience.

**Comment by bdougie** [comment]
**I pushed a few updates just to make this work.**

1. Leveraging persisted queries as a temporary fix until GitHub is able to update the public API permissions.
2. Fixed an oversight in not returning errors in the reducer.
3. Adding button conditional to create a second step, which will be installing the App on the open-sauced goals repo.

**Things that still need to be addressed.**

- [ ] Follow up with GitHub to get a timeline on how long the temporary fix is needed.
- [ ] Updated the reducer to properly handle errors. Consider adding https://github.com/open-sauced/open-sauced/pull/717 then.
- [ ] Check to see if Onegraph wants to create a mutation that installs the GitHub App on the created repo. 

App install in REST: `PUT /user/installations/:installation_id/repositories/:repository_id`

https://developer.github.com/v3/apps/installations/#add-a-repository-to-an-app-installation

**Comment by yg** [comment]
@bdougie 404 on https://github.com/apps/open-sauced/installations/new/permissions?target_id=20134767

File: src/lib/persistedGraphQL.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  return response;
+}
+
+export {
+  persistedDeploymentFetch,
+  persistedIssuesFetch,
+  persistedInteractionsFetch,
+  persistedGoalFetch,
+  persistedRepoDataFetch,
+};
```
Comment (yg) [inline src/lib/persistedGraphQL.js:79]: @bdougie We are not rate-limited with this? Is this free?

https://www.loom.com/share/3aaf3863262348c59986b79b7e649f4a

File: src/components/CreateGoals.js [hunk 2]
```diff
@@ -55,9 +58,15 @@ function CreateGoals({user, onRepoCreation}) {
         </SpaceBetween>
       </ContextStyle>
       <br style={{marginTop: 8}} />
-      <Button primary onClick={_handleRepoCreation}>
-        Create your goal workspace
-      </Button>
+      {installNeeded ? (
+        <a referrer="noreferrer" target="_blank" href="https://github.com/apps/open-sauced/installations/new/permissions?target_id=20134767">
```
Comment (yg) [inline src/components/CreateGoals.js:62]: The `user id` needs to be dynamic!

**Comment by yg** [comment]
We also need to redirect back to https://opensauced.pizza after installation

**Comment by bdougie** [comment]
Great feedback. Thanks for trying this out, there are a few things that need to be implemented to make this work. Having a mutation. For the installation would clean up this experience.

I will look into that 404 and a add a redirect.

File: src/components/CreateGoals.js (hunk 2 above)
Comment (bdougie) [reply src/components/CreateGoals.js:62]: Yep!

File: src/lib/persistedGraphQL.js (hunk 1 above)
Comment (bdougie) [reply src/lib/persistedGraphQL.js:79]: They are limited but it is coming from my app token and cached at an hour.

This is not a perfect solution but hopefully just a temporary one to solve the current permissions mismatch

**Comment by bdougie** [comment]
@yg, when you have a moment. Could you check out this deploy preview? I have completed the login flow which has a new screen that asks the user to install the new Open Sauced App on the goals repo. 

Screenshot 2020-07-05 00 09 22

**Comment by yg** [comment]
@bdougie, After Installing in GitHub it is not redirecting to opensauced.pizza or to the deploy preview

And the Illustration is expanded vertically!

screencapture-deploy-preview-699-open-sauced-netlify-app-2020-07-05-12_42_22

**Comment by bdougie** [comment]
Looks like I cannot add a redirect since auth is already being requested.

Screenshot 2020-07-05 00 30 28

There are some things I need to address, one being checking in with one graph to see if the installation endpoint can be created as a query so I can remove this separate step.

I also need to check to see if I can send the login to opensauced.pizza/authorize instead of serve.onegraph
 
> - [ ] Follow up with GitHub to get a timeline on how long the temporary fix is needed.
> - [ ]  Check to see if Onegraph wants to create a mutation that installs the GitHub App on the created repo.

**Comment by bdougie** [comment]
https://discordapp.com/channels/714698561081704529/714698561081704532/764234676453113896

@strager drove by in the discord, asking about permissions. Linking for reference. 

After chatting with @sgrove, there is a better way to approach persisted queries now that will be less confusing for future contributors. I plan to implement that when I find time, but. due to the event this Thrusday I won't be able to to get to this until the weekend. 

For those interested in learning about Persisted Queries, check out my talk about it GraphQL Enterpise. 

https://www.youtube.com/watch?v=yr5kSZljBxo&list=PLpURC3VhaQD0QkE1uRsGX7gMayocniv5s&index=35&t=2s

**Comment by bdougie** [comment]
Update: This branch now only asks for access to the repo open sauced creates. open-sauced-goals.

Screen Shot 2020-10-23 at 1 52 28 PM
Screen Shot 2020-10-23 at 1 52 19 PM

**Comment by bdougie** [comment]
OK, this PR will be merged with the expectation that there will be errors in the API. If you @open-sauced/triage team, please try and use the site and open issues if you find any. Thanks for your 👀 on this and help making the sauce flow!

**Comment by jasonericdavis** [comment]
@bdougie I took a look today and everything pretty much went smooth. The only hiccup that I had was when presented with the "Install the GitHub App screen" I thought I had to do it in another tab instead of clicking the button under the picture. I was obviously making it more difficult then it needed to be but there was nothing that indicated that the picture was meant to provide instructions for the next popup and that clicking the button would lead me to that screen. Once I clicked the button everything went smooth.

**Comment by bdougie** [comment]
Thanks for going through this @jasonericdavis. I notice Netlify and Vercel do this well. I think I might copy the way they link to the app for installation.

**Pull request merged**